File: src/dotnet/ApiSdk/MarketConfig.cs
using System.Collections.Frozen;
using System.Collections.Immutable;
using Microsoft.Extensions.Configuration;

namespace ApiSdk;

/// <summary>Markets the SDK can resolve V1 flat-file data sources for.</summary>
public enum Market
{
    EU,
    Nordic,
    UK,
    USA,
    Australia,
}

/// <summary>
/// Resolved V1 flat-file paths for a market/locale, ready to feed into the
/// matching fields of <see cref="DataSources"/> (<c>Voyages</c>, <c>Ships</c>,
/// <c>SourceMarkets</c>). <see cref="DataSources.CabinGrades"/> and
/// <see cref="DataSources.Ports"/> are not locale-specific and are left for the
/// caller to supply directly.
/// </summary>
public sealed record MarketDataSources
{
    /// <summary>The locale actually resolved to (lowercase, e.g. "de").</summary>
    public required string Locale { get; init; }

    /// <summary>The settlement currency for <see cref="Locale"/> (uppercase ISO 4217, e.g. "EUR").</summary>
    public required string Currency { get; init; }

    public required string Voyages { get; init; }
    public required string Ships { get; init; }

    /// <summary>Always exactly one file: the source-market rates for <see cref="Currency"/>.</summary>
    public required IReadOnlyList<string> SourceMarkets { get; init; }
}

/// <summary>
/// Resolved V3 flat-file paths for a market/locale, ready to feed into the
/// matching fields of <see cref="DataSources"/> (<c>Voyages</c>, <c>Ships</c>).
/// V3 has no source-market rate files (pricing is embedded per voyage) and no
/// separate cabin-grade reference, so unlike <see cref="MarketDataSources"/>
/// there is no <c>Currency</c>/<c>SourceMarkets</c> here.
/// <see cref="DataSources.Ports"/> is a single flat file in V3 too (not
/// locale-specific) and is left for the caller to supply directly.
/// </summary>
public sealed record MarketDataSourcesV3
{
    /// <summary>The locale actually resolved to (uppercase 2-letter code, e.g. "DE").</summary>
    public required string Locale { get; init; }

    public required string Voyages { get; init; }
    public required string Ships { get; init; }
}

/// <summary>
/// Resolves a <see cref="Market"/> (and optional locale) into the
/// locale-suffixed V1 flat-file paths
/// (<c>voyages_{locale}.json</c>, <c>ships_{locale}.json</c>,
/// <c>SourceMarket_{CURRENCY}_seaware.json</c>), and — mirroring
/// <see cref="DataSourceFormatConfig"/> — reads the selected market/locale from
/// configuration/environment instead of a compiled-in default.
///
/// V3 ("prod") uses a genuinely different locale set — different available
/// countries, different casing/codes (e.g. "GB" not "uk"), and a different
/// Canada/Switzerland grouping decision — so it gets its own internal lookup
/// table and resolver (<see cref="ResolveMarketDataSourcesV3"/>) rather than
/// sharing the V1 one (<see cref="ResolveMarketDataSources"/>).
/// <see cref="ResolveMarket"/>/<see cref="ResolveLocale"/> are format-agnostic
/// and are reused as-is by both.
///
/// The two locale lookup tables and the currency table are deliberately
/// PRIVATE — see the remarks on <see cref="MarketLocales"/> for why — with
/// <see cref="GetLocales"/>, <see cref="TryNormalizeLocale"/> and
/// <see cref="GetCurrency"/> as the only way anything outside this file can
/// observe their contents.
/// </summary>
public static class MarketConfig
{
    /// <summary>The IConfiguration key the market is read from.</summary>
    public const string MarketConfigKey = "DataSources:Market";

    /// <summary>The environment variable the market falls back to.</summary>
    public const string MarketEnvVar = "DATASOURCE_MARKET";

    /// <summary>The IConfiguration key the locale override is read from.</summary>
    public const string LocaleConfigKey = "DataSources:Locale";

    /// <summary>The environment variable the locale override falls back to.</summary>
    public const string LocaleEnvVar = "DATASOURCE_LOCALE";

    /// <summary>
    /// The locales available for each market, lowercase (V1/dev).
    ///
    /// PRIVATE: no public collection type in .NET is truly tamper-proof against
    /// reflection/marshal-level access (there is always some sanctioned BCL
    /// accessor that reaches a collection's backing store — that's not a bug in
    /// those APIs, and reaching it isn't crossing a real security boundary;
    /// .NET has no partial-trust sandboxing), so the only fix that actually
    /// holds is not exposing the collection at all. <see cref="GetLocales"/>/
    /// <see cref="TryNormalizeLocale"/>/<see cref="GetCurrency"/> are the entire
    /// external contract; <see cref="FrozenDictionary{TKey,TValue}"/>/
    /// <see cref="FrozenSet{T}"/> here are for O(1) lookup performance, not
    /// defensive copy-proofing — they don't need to be copy-proof once private.
    /// </summary>
    private static readonly FrozenDictionary<Market, FrozenSet<string>> MarketLocales = new Dictionary<Market, FrozenSet<string>>
    {
        [Market.EU] = new[] { "de", "fr", "dk" }.ToFrozenSet(),
        [Market.Nordic] = new[] { "se", "no" }.ToFrozenSet(),
        [Market.UK] = new[] { "uk" }.ToFrozenSet(),
        [Market.USA] = new[] { "us" }.ToFrozenSet(),
        [Market.Australia] = new[] { "au" }.ToFrozenSet(),
    }.ToFrozenDictionary();

    /// <summary>
    /// The settlement currency (uppercase ISO 4217) for each V1/dev locale.
    /// PRIVATE for the same reason as <see cref="MarketLocales"/> — reachable
    /// only via <see cref="GetCurrency"/>.
    ///
    /// This is V1-ONLY despite this class also modeling V3: a V3 locale code
    /// that happens to share spelling with a V1 one by coincidence of
    /// geography (e.g. "DE") is not guaranteed to resolve the "same" entry on
    /// purpose, and V3-only codes ("GB", "CH", "CA") aren't in here at all —
    /// V3 has no currency concept, see <see cref="MarketDataSourcesV3"/>'s
    /// remarks (pricing is embedded per voyage). The case-insensitive keying
    /// is purely so <see cref="GetCurrency"/> doesn't care about the caller's
    /// casing of a V1 code; it is not a claim that this table also serves V3.
    /// </summary>
    private static readonly FrozenDictionary<string, string> LocaleCurrency = new Dictionary<string, string>
    {
        ["de"] = "EUR",
        ["fr"] = "EUR",
        ["dk"] = "DKK",
        ["se"] = "SEK",
        ["no"] = "NOK",
        ["uk"] = "GBP",
        ["us"] = "USD",
        ["au"] = "AUD",
    }.ToFrozenDictionary(StringComparer.OrdinalIgnoreCase);

    /// <summary>
    /// The locales available for each market in the V3 ("prod") flat-file format,
    /// as their real on-disk uppercase 2-letter codes. A DIFFERENT set from
    /// <see cref="MarketLocales"/> (V1/dev): prod covers 10 countries vs. dev's
    /// 8, "UK" is coded "GB", and business-decision groupings that don't exist
    /// in dev apply here — Canada groups under USA, Switzerland groups under
    /// EU. PRIVATE for the same reason as <see cref="MarketLocales"/>.
    /// </summary>
    private static readonly FrozenDictionary<Market, FrozenSet<string>> MarketLocalesV3 = new Dictionary<Market, FrozenSet<string>>
    {
        [Market.EU] = new[] { "DE", "FR", "DK", "CH" }.ToFrozenSet(),
        [Market.Nordic] = new[] { "SE", "NO" }.ToFrozenSet(),
        [Market.UK] = new[] { "GB" }.ToFrozenSet(),
        [Market.USA] = new[] { "US", "CA" }.ToFrozenSet(),
        [Market.Australia] = new[] { "AU" }.ToFrozenSet(),
    }.ToFrozenDictionary();

    /// <summary>
    /// Resolve the market from <paramref name="config"/> (key
    /// <see cref="MarketConfigKey"/>), falling back to the <see cref="MarketEnvVar"/>
    /// environment variable. Throws <see cref="InvalidOperationException"/> when
    /// neither is set, and <see cref="InvalidOperationException"/> when the value
    /// is present but not a valid <see cref="Market"/> name.
    /// </summary>
    public static Market ResolveMarket(IConfiguration? config = null)
    {
        var raw = config?[MarketConfigKey];
        if (string.IsNullOrWhiteSpace(raw))
            raw = Environment.GetEnvironmentVariable(MarketEnvVar);

        if (string.IsNullOrWhiteSpace(raw))
            throw new InvalidOperationException(
                $"Market is not configured. Set the '{MarketConfigKey}' configuration " +
                $"key or the '{MarketEnvVar}' environment variable to one of: " +
                $"{string.Join(", ", Enum.GetNames<Market>())}. There is no compiled-in default.");

        if (!TryParseEnumNameStrict<Market>(raw.Trim(), out var market))
        {
            throw new InvalidOperationException(
                $"Invalid market '{raw}'. Expected one of: {string.Join(", ", Enum.GetNames<Market>())} " +
                $"(from '{MarketConfigKey}' or '{MarketEnvVar}').");
        }

        return market;
    }

    /// <summary>
    /// Case-insensitively match <paramref name="raw"/> against
    /// <typeparamref name="TEnum"/>'s actual member NAMES only.
    ///
    /// <see cref="Enum.TryParse{TEnum}(string, bool, out TEnum)"/> combined with
    /// <see cref="Enum.IsDefined(Type, object)"/> is NOT sufficient input
    /// validation on its own: <c>TryParse</c> also accepts the underlying
    /// numeric value as a string (so <c>"3"</c> parses to whatever member has
    /// value 3), and for comma-separated input it happily ORs the parsed values
    /// together even when the enum isn't <c>[Flags]</c> — e.g. for
    /// <see cref="Market"/> (EU=0, Nordic=1, UK=2, USA=3, Australia=4),
    /// <c>"Nordic,UK"</c> parses to 1|2=3, and <c>Enum.IsDefined(3)</c> then
    /// reports "defined" purely because 3 happens to also be USA's value — so
    /// an operator typo silently resolves to the wrong market. Matching against
    /// <see cref="Enum.GetNames{TEnum}"/> directly closes both holes.
    /// </summary>
    private static bool TryParseEnumNameStrict<TEnum>(string raw, out TEnum value) where TEnum : struct, Enum
    {
        foreach (var name in Enum.GetNames<TEnum>())
        {
            if (string.Equals(name, raw, StringComparison.OrdinalIgnoreCase))
            {
                value = Enum.Parse<TEnum>(name); // exact-case parse of a known-good member name
                return true;
            }
        }

        value = default;
        return false;
    }

    /// <summary>
    /// Resolve the locale override from <paramref name="config"/> (key
    /// <see cref="LocaleConfigKey"/>), falling back to the <see cref="LocaleEnvVar"/>
    /// environment variable. Unlike <see cref="ResolveMarket"/> this does NOT throw
    /// when unset — a locale is only required for markets with more than one
    /// locale, and <see cref="ResolveMarketDataSources"/> is what enforces that.
    /// </summary>
    public static string? ResolveLocale(IConfiguration? config = null)
    {
        var raw = config?[LocaleConfigKey];
        if (string.IsNullOrWhiteSpace(raw))
            raw = Environment.GetEnvironmentVariable(LocaleEnvVar);

        return string.IsNullOrWhiteSpace(raw) ? null : raw.Trim();
    }

    /// <summary>
    /// The locales available for <paramref name="market"/> under <paramref name="format"/>,
    /// as a fresh, alphabetically ordered, independently-owned snapshot — the
    /// one supported way to enumerate them from outside this file (e.g. for a
    /// UI picker; see <c>ApiSdk.SDKCLI/Program.cs</c>'s locale menu). A set has
    /// no ordering guarantee, so sorting is done here rather than leaving every
    /// caller to reinvent "what order do I show these in". Throws
    /// <see cref="InvalidOperationException"/> if <paramref name="market"/> has
    /// no entry for <paramref name="format"/> (a <see cref="Market"/> member
    /// added without updating the matching table — a bug, not a runtime input
    /// to handle gracefully).
    /// </summary>
    public static IReadOnlyList<string> GetLocales(Market market, DataSourceFormat format)
    {
        var table = format == DataSourceFormat.V3 ? MarketLocalesV3 : MarketLocales;
        if (!table.TryGetValue(market, out var locales))
        {
            throw new InvalidOperationException(
                $"No locale table entry for market '{market}' under format '{format}'. This means a " +
                $"{nameof(Market)} member was added without a matching table entry.");
        }

        return locales.OrderBy(l => l, StringComparer.Ordinal).ToArray();
    }

    /// <summary>
    /// Normalize <paramref name="locale"/> to the casing <paramref name="format"/>
    /// expects (V1 lowercase, V3 uppercase) and confirm it's valid for
    /// <paramref name="market"/>, in one step. This is the ONE place that
    /// decides V1-lowercase vs. V3-uppercase — callers (e.g. the TUI's locale
    /// prompt in <c>ApiSdk.SDKCLI/Program.cs</c>) must route through this
    /// rather than re-deriving the casing rule themselves. Returns
    /// <see langword="false"/> (with <paramref name="normalized"/>
    /// <see langword="null"/>) if <paramref name="locale"/> is blank or not
    /// valid for this market+format — this supersedes what used to be a
    /// separate <c>IsValidLocale</c> bool-only check, which had no production
    /// caller and forced everyone who actually needed the normalized string
    /// (i.e. every real caller) to re-derive it anyway.
    /// </summary>
    public static bool TryNormalizeLocale(Market market, string? locale, DataSourceFormat format, out string? normalized)
    {
        if (string.IsNullOrWhiteSpace(locale))
        {
            normalized = null;
            return false;
        }

        var candidate = NormalizeLocale(locale, format);
        var table = format == DataSourceFormat.V3 ? MarketLocalesV3 : MarketLocales;
        if (table.TryGetValue(market, out var locales) && locales.Contains(candidate))
        {
            normalized = candidate;
            return true;
        }

        normalized = null;
        return false;
    }

    /// <summary>
    /// The settlement currency for a V1/dev <paramref name="locale"/> (any
    /// case, whitespace trimmed). Throws <see cref="InvalidOperationException"/> —
    /// not <see cref="ArgumentNullException"/>/<see cref="ArgumentException"/> —
    /// for a null/blank/unrecognized locale alike, matching every other
    /// "not found" case in this class rather than giving null a different
    /// failure mode than "just wrong". There is no V3 equivalent of this
    /// lookup (see <see cref="LocaleCurrency"/>'s remarks).
    /// </summary>
    public static string GetCurrency(string? locale)
    {
        if (!string.IsNullOrWhiteSpace(locale) && LocaleCurrency.TryGetValue(locale.Trim(), out var currency))
            return currency;

        throw new InvalidOperationException($"No currency mapping for locale '{locale}'.");
    }

    /// <summary>V1 locales are stored/compared lowercase, V3 uppercase.</summary>
    private static string NormalizeLocale(string locale, DataSourceFormat format) =>
        format == DataSourceFormat.V3 ? locale.Trim().ToUpperInvariant() : locale.Trim().ToLowerInvariant();

    /// <summary>
    /// Resolve the voyages/ships/source-market flat-file paths for
    /// <paramref name="market"/> under <paramref name="baseDir"/>.
    ///
    /// If the market has more than one locale, <paramref name="locale"/> is
    /// REQUIRED — a null/blank locale throws, there is no silent "pick the first
    /// locale" default. If the market has exactly one locale, <paramref name="locale"/>
    /// defaults to it when null/blank. A locale that doesn't belong to the given
    /// market (whichever way it was resolved) also throws.
    /// </summary>
    public static MarketDataSources ResolveMarketDataSources(Market market, string? locale, string baseDir)
    {
        if (string.IsNullOrWhiteSpace(baseDir))
            throw new ArgumentException("Base directory must be provided.", nameof(baseDir));

        if (!MarketLocales.TryGetValue(market, out var locales))
            throw new InvalidOperationException(
                $"No V1 locale table entry for market '{market}'. This means a {nameof(Market)} " +
                $"member was added without a matching locale table entry.");

        string resolvedLocale;
        if (string.IsNullOrWhiteSpace(locale))
        {
            if (locales.Count != 1)
            {
                throw new InvalidOperationException(
                    $"Market '{market}' has multiple locales ({FormatLocales(locales)}); " +
                    "a locale must be specified explicitly. There is no silent default.");
            }
            resolvedLocale = locales.Single();
        }
        else
        {
            resolvedLocale = NormalizeLocale(locale, DataSourceFormat.V1);
            if (!locales.Contains(resolvedLocale))
            {
                throw new InvalidOperationException(
                    $"Locale '{locale}' is not valid for market '{market}'. Expected one of: " +
                    $"{FormatLocales(locales)}.");
            }
        }

        var currency = GetCurrency(resolvedLocale);

        return new MarketDataSources
        {
            Locale = resolvedLocale,
            Currency = currency,
            Voyages = Path.Combine(baseDir, $"voyages_{resolvedLocale}.json"),
            Ships = Path.Combine(baseDir, $"ships_{resolvedLocale}.json"),
            // ImmutableArray.Create, not `new[] {...}` behind IReadOnlyList<string>
            // — `new[] {...}` is exactly the "looks read-only, isn't" pattern
            // this file otherwise avoids. Unlike the module-level tables above,
            // this is a freshly-allocated, non-static, per-call value: there is
            // no shared state for a cast-and-mutate (or marshal-and-mutate) to
            // corrupt even in principle, so ImmutableArray's guarantees are
            // exactly as strong here as they're ever going to get — the
            // remaining gap only mattered for the process-static tables.
            SourceMarkets = ImmutableArray.Create(Path.Combine(baseDir, $"SourceMarket_{currency}_seaware.json")),
        };
    }

    /// <summary>Sorted, comma-joined locales for error messages — deterministic output, not iteration order.</summary>
    private static string FormatLocales(FrozenSet<string> locales) =>
        string.Join(", ", locales.OrderBy(l => l, StringComparer.Ordinal));

    /// <summary>
    /// Resolve the voyages/ships flat-file paths for <paramref name="market"/>
    /// under <paramref name="baseDir"/>, using the V3 ("prod") locale set —
    /// uppercase 2-letter codes, no source-market/currency concept.
    ///
    /// Same "explicit required" convention as <see cref="ResolveMarketDataSources"/>:
    /// if the market has more than one V3 locale, <paramref name="locale"/> is
    /// REQUIRED — a null/blank locale throws, there is no silent "pick the first
    /// locale" default. If the market has exactly one V3 locale, <paramref name="locale"/>
    /// defaults to it when null/blank. A locale that doesn't belong to the given
    /// market (whichever way it was resolved) also throws.
    /// </summary>
    public static MarketDataSourcesV3 ResolveMarketDataSourcesV3(Market market, string? locale, string baseDir)
    {
        if (string.IsNullOrWhiteSpace(baseDir))
            throw new ArgumentException("Base directory must be provided.", nameof(baseDir));

        if (!MarketLocalesV3.TryGetValue(market, out var locales))
            throw new InvalidOperationException(
                $"No V3 locale table entry for market '{market}'. This means a {nameof(Market)} " +
                $"member was added without a matching locale table entry.");

        string resolvedLocale;
        if (string.IsNullOrWhiteSpace(locale))
        {
            if (locales.Count != 1)
            {
                throw new InvalidOperationException(
                    $"Market '{market}' has multiple V3 locales ({FormatLocales(locales)}); " +
                    "a locale must be specified explicitly. There is no silent default.");
            }
            resolvedLocale = locales.Single();
        }
        else
        {
            resolvedLocale = NormalizeLocale(locale, DataSourceFormat.V3);
            if (!locales.Contains(resolvedLocale))
            {
                throw new InvalidOperationException(
                    $"Locale '{locale}' is not valid for V3 market '{market}'. Expected one of: " +
                    $"{FormatLocales(locales)}.");
            }
        }

        return new MarketDataSourcesV3
        {
            Locale = resolvedLocale,
            Voyages = Path.Combine(baseDir, $"voyages_{resolvedLocale}.json"),
            Ships = Path.Combine(baseDir, $"ships_{resolvedLocale}.json"),
        };
    }
}

File: src/js/src/loading/marketConfig.ts
import * as path from 'path';

/**
 * Selects which market's flat-file sample data {@link resolveMarketDataSources}
 * resolves paths for. Each market maps to one or more locale-suffixed file sets
 * under RefData (see {@link MARKET_LOCALES}).
 */
export type Market = 'EU' | 'Nordic' | 'UK' | 'USA' | 'Australia';

/**
 * The locale codes that belong to each {@link Market}, in the order the sample
 * data was generated. A market with more than one locale has no "default" —
 * the caller must say which one they want (see {@link resolveMarketDataSources}).
 */
export const MARKET_LOCALES: Record<Market, readonly string[]> = {
  EU: ['de', 'fr', 'dk'],
  Nordic: ['se', 'no'],
  UK: ['uk'],
  USA: ['us'],
  Australia: ['au'],
};

/**
 * The settlement currency for each locale, used to pick the matching
 * `SourceMarket_{currency}_seaware.json` rate file.
 */
export const LOCALE_CURRENCY: Record<string, string> = {
  de: 'EUR',
  fr: 'EUR',
  dk: 'DKK',
  se: 'SEK',
  no: 'NOK',
  uk: 'GBP',
  us: 'USD',
  au: 'AUD',
};

const MARKETS = Object.keys(MARKET_LOCALES) as Market[];

/**
 * The environment variable that carries the market selection. Mirrors
 * {@link DATASOURCE_FORMAT_ENV} in {@link ./formatConfig}: the value lives in
 * configuration (process environment), not compiled into the SDK.
 */
export const DATASOURCE_MARKET_ENV = 'DATASOURCE_MARKET';

/**
 * The environment variable that carries the locale selection. Only required
 * when the resolved {@link Market} has more than one locale — see
 * {@link resolveMarketDataSources}.
 */
export const DATASOURCE_LOCALE_ENV = 'DATASOURCE_LOCALE';

/**
 * Resolve the {@link Market} from configuration.
 *
 * The value is sourced from the `DATASOURCE_MARKET` environment variable
 * (case-insensitive match against {@link MARKET_LOCALES}' keys). There is NO
 * silent default: if the variable is unset, blank, or holds an unrecognized
 * value, this THROWS a clear Error. Mirrors {@link resolveDataSourceFormat}'s
 * throw-on-missing semantics.
 *
 * @param env The environment to read from (defaults to `process.env`); injectable for tests.
 * @throws {Error} when `DATASOURCE_MARKET` is unset/blank or not a recognized market.
 */
export function resolveMarket(env: NodeJS.ProcessEnv = process.env): Market {
  const raw = env[DATASOURCE_MARKET_ENV];

  if (raw === undefined || raw.trim().length === 0) {
    throw new Error(
      `${DATASOURCE_MARKET_ENV} is not set. It must be one of ${MARKETS.join(', ')} — there is no default.`
    );
  }

  const trimmed = raw.trim();
  const match = MARKETS.find((m) => m.toLowerCase() === trimmed.toLowerCase());
  if (match) {
    return match;
  }

  throw new Error(
    `${DATASOURCE_MARKET_ENV} has an unrecognized value "${raw}". ` +
      `Expected one of ${MARKETS.join(', ')}.`
  );
}

/**
 * Resolve the locale from configuration.
 *
 * The value is sourced from the `DATASOURCE_LOCALE` environment variable.
 * Unlike {@link resolveMarket}, this does NOT throw when unset — whether a
 * locale is required depends on the resolved market (see
 * {@link resolveMarketDataSources}), so that check happens there instead.
 *
 * @param env The environment to read from (defaults to `process.env`); injectable for tests.
 * @returns The trimmed, lower-cased locale, or `undefined` if unset/blank.
 */
export function resolveLocale(env: NodeJS.ProcessEnv = process.env): string | undefined {
  const raw = env[DATASOURCE_LOCALE_ENV];
  if (raw === undefined || raw.trim().length === 0) {
    return undefined;
  }
  return raw.trim().toLowerCase();
}

/** The subset of {@link DataSources} paths that vary by market/locale. */
export interface MarketDataSources {
  voyages: string;
  ships: string;
  sourceMarkets: string[];
}

/**
 * Resolve the market/locale-specific flat-file paths for {@link DataSources}.
 *
 * If {@link market} has exactly one locale, `locale` is optional and defaults
 * to it. If {@link market} has more than one locale, `locale` is REQUIRED and
 * MUST be one of that market's locales — there is no default. Both cases THROW
 * a clear Error on invalid input, mirroring {@link resolveDataSourceFormat}'s
 * no-silent-fallback convention.
 *
 * @param market The resolved {@link Market} (see {@link resolveMarket}).
 * @param locale The resolved locale (see {@link resolveLocale}), or `undefined`.
 * @param baseDir Absolute path to the RefData directory the files live under.
 * @throws {Error} when `locale` is missing but required, or not valid for `market`.
 */
export function resolveMarketDataSources(
  market: Market,
  locale: string | undefined,
  baseDir: string
): MarketDataSources {
  const locales = MARKET_LOCALES[market];

  let resolvedLocale: string;
  if (locales.length === 1) {
    resolvedLocale = locale?.trim() ? locale.trim().toLowerCase() : locales[0];
  } else {
    if (locale === undefined || locale.trim().length === 0) {
      throw new Error(
        `${DATASOURCE_LOCALE_ENV} is not set. Market "${market}" has multiple locales ` +
          `(${locales.join(', ')}) and requires an explicit locale — there is no default.`
      );
    }
    resolvedLocale = locale.trim().toLowerCase();
  }

  if (!locales.includes(resolvedLocale)) {
    throw new Error(
      `Locale "${locale}" is not valid for market "${market}". ` +
        `Expected one of: ${locales.join(', ')}.`
    );
  }

  const currency = LOCALE_CURRENCY[resolvedLocale];
  if (!currency) {
    throw new Error(`No currency configured for locale "${resolvedLocale}".`);
  }

  return {
    voyages: path.join(baseDir, `voyages_${resolvedLocale}.json`),
    ships: path.join(baseDir, `ships_${resolvedLocale}.json`),
    sourceMarkets: [path.join(baseDir, `SourceMarket_${currency}_seaware.json`)],
  };
}

// =============================================================================
// V3 (prod) — a SEPARATE table, deliberately not merged with the V1 one above.
//
// Prod (data/flatfiles_prod) uses a different naming scheme entirely: uppercase
// 2-letter country codes (no `_seaware` suffix), no SourceMarket rate files
// (v3 pricing is embedded per voyage — see V3DataSetLoader), and no locale-
// scoped ports/cabin-grades file. V1 dev and V3 prod also cover genuinely
// different, only-partially-overlapping locale sets (dev has no CA/CH
// fixtures; prod has no `uk` — it's `GB`), so they get independent tables
// rather than one shared shape.
// =============================================================================

/**
 * The V3 (prod) locale codes for each {@link Market}, already upper-cased —
 * that is the real on-disk casing (`voyages_AU.json`, not `voyages_au.json`).
 *
 * Business grouping differs from {@link MARKET_LOCALES} (V1/dev) by design:
 * Canada (`CA`) groups under USA, Switzerland (`CH`) groups under EU — dev has
 * no fixtures for either, so this is a prod-only distinction.
 */
export const MARKET_LOCALES_V3: Record<Market, readonly string[]> = {
  EU: ['DE', 'FR', 'DK', 'CH'],
  Nordic: ['SE', 'NO'],
  UK: ['GB'],
  USA: ['US', 'CA'],
  Australia: ['AU'],
};

/** The subset of {@link DataSources} paths V3 (prod) varies by market/locale. */
export interface MarketDataSourcesV3 {
  voyages: string;
  ships: string;
}

/**
 * Resolve the V3 (prod) market/locale-specific flat-file paths for
 * {@link DataSources}.
 *
 * Same convention as {@link resolveMarketDataSources}: if {@link market} has
 * exactly one locale, `locale` is optional and defaults to it; if it has more
 * than one, `locale` is REQUIRED and MUST be one of that market's locales —
 * no default, THROWS a clear Error otherwise. `locale` is matched
 * case-insensitively (env values are lower-cased by {@link resolveLocale}) but
 * resolved paths always use the upper-case on-disk codes from
 * {@link MARKET_LOCALES_V3}.
 *
 * Unlike the V1 resolver, there is no currency/`sourceMarkets` concept in
 * prod — pricing is embedded per voyage — so only `voyages`/`ships` are
 * returned; `ports` is flat and un-suffixed (not market/locale-specific),
 * same as V1's `cabinGrades`/`ports`, so callers build that path separately.
 *
 * @param market The resolved {@link Market} (see {@link resolveMarket}).
 * @param locale The resolved locale (see {@link resolveLocale}), or `undefined`.
 * @param baseDir Absolute path to the prod flat-files directory.
 * @throws {Error} when `locale` is missing but required, or not valid for `market`.
 */
export function resolveMarketDataSourcesV3(
  market: Market,
  locale: string | undefined,
  baseDir: string
): MarketDataSourcesV3 {
  const locales = MARKET_LOCALES_V3[market];

  let resolvedLocale: string;
  if (locales.length === 1) {
    resolvedLocale = locale?.trim() ? locale.trim().toUpperCase() : locales[0];
  } else {
    if (locale === undefined || locale.trim().length === 0) {
      throw new Error(
        `${DATASOURCE_LOCALE_ENV} is not set. Market "${market}" has multiple locales ` +
          `(${locales.join(', ')}) and requires an explicit locale — there is no default.`
      );
    }
    resolvedLocale = locale.trim().toUpperCase();
  }

  if (!locales.includes(resolvedLocale)) {
    throw new Error(
      `Locale "${locale}" is not valid for market "${market}". ` +
        `Expected one of: ${locales.join(', ')}.`
    );
  }

  return {
    voyages: path.join(baseDir, `voyages_${resolvedLocale}.json`),
    ships: path.join(baseDir, `ships_${resolvedLocale}.json`),
  };
}
